### ISSUE TITLE
Building TF Lite C++ shared library for macOS, linux, iOS and Android

### ISSUE THREAD

**Issue by ebraraktas** [tensorflow/tensorflow#40438]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): macOS Mojave (10.14)
- TensorFlow installed from (source or binary): source
- TensorFlow version: r2.2
- Python version: 3.7
- Installed using virtualenv? pip? conda?: pip
- Bazel version (if compiling from source): 3.1.0
- GCC/Compiler version (if compiling from source): Apple clang version 11.0.0 (clang-1100.0.33.17)
- CUDA/cuDNN version: -
- GPU model and memory: -

I am trying to develop a C++ library on macOS and ubuntu to use in Android and iOS. Therefore, I am trying to build TFLite C++ shared library to link it my own C++ shared library using CMake. I have tried to build that using command `bazel build -c opt //tensorflow/lite:tensorflowlite` and got `libtensorflowlite.dylib` file as bazel-output. However CMake fails to link it and gives the following error:
```
Undefined symbols for architecture x86_64:
  "tflite::FlatBufferModel::BuildFromFile(char const*, tflite::ErrorReporter*)", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "tflite::FlatBufferModel::~FlatBufferModel()", referenced from:
      std::__1::default_delete<tflite::FlatBufferModel>::operator()(tflite::FlatBufferModel*) const in my_class.cpp.o
  "tflite::DefaultErrorReporter()", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "tflite::ops::builtin::BuiltinOpResolver::BuiltinOpResolver()", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "tflite::impl::Interpreter::AllocateTensors()", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
      MyClass::run_model(std::__1::vector<std::__1::vector<float, std::__1::allocator<float> >, std::__1::allocator<std::__1::vector<float, std::__1::allocator<float> > > > const&) in my_class.cpp.o
  "tflite::impl::Interpreter::ResizeInputTensor(int, std::__1::vector<int, std::__1::allocator<int> > const&)", referenced from:
      MyClass::run_model(std::__1::vector<std::__1::vector<float, std::__1::allocator<float> >, std::__1::allocator<std::__1::vector<float, std::__1::allocator<float> > > > const&) in my_class.cpp.o
  "tflite::impl::Interpreter::Invoke()", referenced from:
      MyClass::run_model(std::__1::vector<std::__1::vector<float, std::__1::allocator<float> >, std::__1::allocator<std::__1::vector<float, std::__1::allocator<float> > > > const&) in my_class.cpp.o
  "tflite::impl::Interpreter::~Interpreter()", referenced from:
      std::__1::default_delete<tflite::impl::Interpreter>::operator()(tflite::impl::Interpreter*) const in my_class.cpp.o
  "tflite::impl::InterpreterBuilder::InterpreterBuilder(tflite::FlatBufferModel const&, tflite::OpResolver const&)", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "tflite::impl::InterpreterBuilder::~InterpreterBuilder()", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "tflite::impl::InterpreterBuilder::operator()(std::__1::unique_ptr<tflite::impl::Interpreter, std::__1::default_delete<tflite::impl::Interpreter> >*)", referenced from:
      MyClass::MyClass(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, int, int) in my_class.cpp.o
  "vtable for tflite::MutableOpResolver", referenced from:
      tflite::MutableOpResolver::~MutableOpResolver() in my_class.cpp.o
```
But, If I try to link shared tflite library(`libtensorflowlite.dylib`) with my static library, it builds without error.

I am fairly new to `bazel` and CMake, however I thought it is caused by building the `libtensorflowlite.dylib` with missing x86_64 architecture and tried below commands and got following errors.:
```
bazel build -c opt --config=ios_fat  //tensorflow/lite:tensorflowlite
ERROR: /Users/username/tensorflow/tensorflow/lite/kernels/internal/BUILD:687:1: C++ compilation of rule '//tensorflow/lite/kernels/internal:kernel_utils' failed (Exit 1)
clang: error: invalid version number in '-mmacosx-version-min=13.2'
```
and 
```
bazel build -c opt --cpu=x86_64  //tensorflow/lite:tensorflowlite
ERROR: /private/var/tmp/_bazel_username/ec3ee607500668f8e05b1977bf58c2f5/external/local_config_cc/BUILD:41:1: in cc_toolchain_suite rule @local_config_cc//:toolchain: cc_toolchain_suite '@local_config_cc//:toolchain' does not contain a toolchain for cpu 'x86_64'
ERROR: Analysis of target '//tensorflow/lite:tensorflowlite' failed; build aborted: Analysis of target '@local_config_cc//:toolchain' failed
```

Finally, I thought it is related to #36178 , #39876 and #35386, but couldn't resolve it.

**Comment by ebraraktas** [comment]
As I said, I was (and still am) new to CMake. And CMakeList.txt of my project was missing some link parameters. I added them and it is working. Therefore, there was not an issue at all. Sorry to bother.

**Comment by wisonye** [comment]
> And CMakeList.txt of my project was missing some link parameters. I added them and it is working. Therefore, there was not an issue at all. Sorry to bother.

@ebraraktas Hi there, could u list all the details how you solve it plz? Definitely need your help on this:)

**Comment by ebraraktas** [comment]
@wisonye 
```
ADD_LIBRARY(tensorflowlite SHARED IMPORTED)
set_property(TARGET tensorflowlite PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_SOURCE_DIR}/lib/libtensorflowlite.dylib)
```
The lines I have added to solve the issue are given above. As you may guess, first you need to build `tensorflowlite` shared library with bazel and copy it to `${CMAKE_CURRENT_SOURCE_DIR}/lib`. Then you can link your cmake target with the line below:
```
target_link_libraries(my_awesome_target tensorflowlite)
```

**Comment by alejouribesanchez** [comment]
@ebraraktas could you show me the whole CMakeLists.txt? I tried your code above but I am still getting "Undefined symbols for architecture x86_64"

This is my CMakeLists.txt

`cmake_minimum_required(VERSION 3.17)
project(veryfi_cpp)
set(CMAKE_CXX_STANDARD 11)
#include directories
include_directories(includes/tensorflow)
include_directories(includes/flatbuffers)

# OPENCV
find_package(OpenCV REQUIRED)
include_directories(${OpenCV_INCLUDE_DIRS})
message("OpenCV_INCLUDE_DIRS = ${OpenCV_INCLUDE_DIRS}")
message("OpenCV_LIBS = ${OpenCV_LIBS}")

# TensorFlowLite
add_library(tensorflowlite SHARED IMPORTED)
set_property(TARGET tensorflowlite PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_SOURCE_DIR}/lib/libtensorflowlite.dylib)
add_executable(veryfi_cpp main.cpp)

# linking libraries
target_link_libraries(veryfi_cpp ${OpenCV_LIBS} tensorflowlite)`

**Comment by ebraraktas** [comment]
@alejouribesanchez As I wrote in my first comment, I was using `bazel build -c opt //tensorflow/lite:tensorflowlite` (You may add some optimization parameters for your target, I recommend it, use `--config=opt` at least). I have added my `CMakeLists.txt` below.

```CMake
cmake_minimum_required(VERSION 3.15)
project(TFLiteExample)

set(CMAKE_CXX_STANDARD 14)

# include has 2 subdirectories: tensorflow and flatbuffers
INCLUDE_DIRECTORIES(${CMAKE_CURRENT_SOURCE_DIR}/include)

# lib has 1 file: libtensorflowlite.dylib
ADD_LIBRARY(tensorflowlite SHARED IMPORTED)
set_property(TARGET tensorflowlite PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_SOURCE_DIR}/lib/libtensorflowlite.dylib)

add_executable(TFLiteExample src/main.cpp)
target_link_libraries(TFLiteExample tensorflowlite)
```

And tree structure of directory:
```
├── CMakeLists.txt
├── README.md
├── include
│   ├── flatbuffers
│   └── tensorflow
├── lib
│   └── libtensorflowlite.dylib
└── src
    └── main.cpp
```

__The important thing is that you should use header files from the branch you build the tensorflow.__ I'm sorry that I was late, but I have discovered this fact while testing my commands before writing this message. __You cannot mix headers and binary from different branches__. I have tested the build command and the `CMakeLists.txt` above with `r2.4` and `r2.3` branches just before I wrote this. Unfortunately, there is no easy way to find and copy necessary headers to your include directory. I had a sloppy Python script at that time but I don't have it right now. 

Hope it helps, please don't hesitate to ask further if you need.

**Comment by alejouribesanchez** [comment]
Thanks, @ebraraktas for your detailed response, I was able to run your example using C++ and flatbuffers 1.12.0 and it worked for me but it seems that the C support is still not working, I mean the C API code, try this main.cpp

#include <iostream>
#include "include/tensorflow/lite/c/c_api.h"

```
int main() {
    const char* model_path = "../models/model.tflite";
    TfLiteModel* model = TfLiteModelCreateFromFile(model_path);
    TfLiteInterpreterOptions* options = TfLiteInterpreterOptionsCreate();
    TfLiteInterpreterOptionsSetNumThreads(options, 3);
    TfLiteInterpreter* interpreter = TfLiteInterpreterCreate(model, options);

// Allocate tensors and populate the input tensor data.
    TfLiteInterpreterAllocateTensors(interpreter);
    TfLiteTensor* inputTensor = TfLiteInterpreterGetInputTensor(interpreter, 0);
    int inputImageHeight = inputTensor->dims->data[1];
    int inputImageWidth = inputTensor->dims->data[2];
    std::cout << "inputImageHeight: " << inputImageHeight << std::endl;
    std::cout << "inputImageWidth: " << inputImageWidth << std::endl;
}
```

And then you will get my first issue reported

```
Undefined symbols for architecture x86_64:
  "_TfLiteInterpreterAllocateTensors", referenced from:
      _main in main.cpp.o
  "_TfLiteInterpreterCreate", referenced from:
      _main in main.cpp.o
  "_TfLiteInterpreterGetInputTensor", referenced from:
      _main in main.cpp.o
  "_TfLiteInterpreterOptionsCreate", referenced from:
      _main in main.cpp.o
  "_TfLiteInterpreterOptionsSetNumThreads", referenced from:
      _main in main.cpp.o
  "_TfLiteModelCreateFromFile", referenced from:
      _main in main.cpp.o
ld: symbol(s) not found for architecture x86_64
```
So let me know if you find a way to run the C code instead of the C++ code.

**Comment by ebraraktas** [comment]
@alejouribesanchez I have built your code with no issue. Are you sure you are linking it with the correct library, to use C API you should build with command below:

```bash
bazel build -c opt --config=opt  //tensorflow/lite/c:tensorflowlite_c
```
And link your target with `libtensorflowlite_c.dylib`, see lines below:
```CMake
ADD_LIBRARY(tensorflowlite_c SHARED IMPORTED)
set_property(TARGET tensorflowlite_c PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_SOURCE_DIR}/lib/libtensorflowlite_c.dylib)

add_executable(TFLiteC main.cpp)
target_link_libraries(TFLiteC tensorflowlite_c)
```

**Comment by alejouribesanchez** [comment]
@ebraraktas thank you so much!, I was using the previous `build -c opt --config=opt  //tensorflow/lite:tensorflowlite` so I was not linking the correct library this `bazel build -c opt --config=opt  //tensorflow/lite/c:tensorflowlite_c` worked with my C code

**Comment by k1ngcyk** [comment]
> @alejouribesanchez I have built your code with no issue. Are you sure you are linking it with the correct library, to use C API you should build with command below:
> 
> ```shell
> bazel build -c opt --config=opt  //tensorflow/lite/c:tensorflowlite_c
> ```
> 
> And link your target with `libtensorflowlite_c.dylib`, see lines below:
> 
> ```cmake
> ADD_LIBRARY(tensorflowlite_c SHARED IMPORTED)
> set_property(TARGET tensorflowlite_c PROPERTY IMPORTED_LOCATION ${CMAKE_CURRENT_SOURCE_DIR}/lib/libtensorflowlite_c.dylib)
> 
> add_executable(TFLiteC main.cpp)
> target_link_libraries(TFLiteC tensorflowlite_c)
> ```

how to build tflite to work with iOS and Android in this way?
using `bazel build -c opt --config=opt //tensorflow/lite:tensorflowlite` I got `Mach-O 64-bit dynamically linked shared library x86_64`
how to get iOS and android arm64?

**Comment by ebraraktas** [comment]
@k1ngcyk I didn't continue my C++ project for cross platform library. Therefore, I cannot suggest a tested solution to your problem. But I think it would be easiest to build TensorFlowLiteC shared library for different platforms and using API of it. Note that this may lack platform specific accelerators like NNAPI, Metal etc. Another solution may be using experimental CMake integration, but it seems not supporting iOS right now. https://www.tensorflow.org/lite/guide/build_cmake#create_a_cmake_project_which_uses_tensorflow_lite

**Comment by fzyzcjy** [comment]
@k1ngcyk @ebraraktas Hi I also need to run tflite with C++ on android+ios+macos+linux. Have you succeeded in doing it - could you please share a bit of hints? (From your latest comments seems that there is no easy way?) Thanks!

**Comment by ShariqM** [comment]
@fzyzcjy see this issue too: https://github.com/tensorflow/tensorflow/issues/43955#issuecomment-757103893

**Comment by ebraraktas** [comment]
Hi @fzyzcjy , 
First of all, I am sorry for late response.

As I stated above I had abandoned the C++ project, and using Rust currently. I have been using [this crate](https://crates.io/crates/tflitec) (disclaimer: I am the developer of it). 

On the other hand, I have tried [this official guide](https://www.tensorflow.org/lite/guide/build_cmake) about CMake build and succeeded to build a toy project for all of the platforms you mentioned. Note that I have used toolchain in [this repository](https://github.com/leetal/ios-cmake) and [this command](https://www.tensorflow.org/lite/guide/build_cmake#cross-compilation) given in the official guide to build it for iOS. Similarly, you need to use the toolchain in the NDK to build it for Android.

The issue about the CMake for me was that I couldn't build stand-alone `TensorFlowLite` shared library. Therefore, the Rust library that I have mentioned about builds `TensorFlowLiteC` shared library, as well. However, you can build your project by adding TensorFlowLite as a subdirectory to your `CMakeLists.txt`, see [this section](https://www.tensorflow.org/lite/guide/build_cmake#create_a_cmake_project_which_uses_tensorflow_lite) please.

**Comment by fzyzcjy** [comment]
@ebraraktas Thanks all the same for the late reply.

Indeed I have also used Rust and wrap the tflite-C API a few days ago. After looking at your code and compare with mine, a small suggestion: Your `Tensor` struct has no lifetime, so `Tensor.tensor_ptr` can live longer than it should. For example, a user can keep hold of a Tensor for a longer time than a Model/Interpreter. Then when he wants to use it, that pointer is not valid anymore. So maybe add a lifetime for Tensor struct ;)

I have compiled on Android and MacOS by using [this readme](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/c/README.md), and ios by [using this](https://www.tensorflow.org/lite/guide/build_ios#build_tensorflowlitec_dynamic_framework_recommended).

**Comment by zhitaop** [comment]
Hi @ebraraktas, sorry for replying to this old thread, but I'm also trying to build the TFLite c++ library on iOS using CMake. 


> On the other hand, I have tried [this official guide](https://www.tensorflow.org/lite/guide/build_cmake) about CMake build and succeeded to build a toy project for all of the platforms you mentioned. Note that I have used toolchain in [this repository](https://github.com/leetal/ios-cmake) and [this command](https://www.tensorflow.org/lite/guide/build_cmake#cross-compilation) given in the official guide to build it for iOS. Similarly, you need to use the toolchain in the NDK to build it for Android.

I have tried building in CMake using the above toolchain and this command (inside a subfolder in my local TensorFlow repo which contains the toolchain file):
`cmake -G Xcode -DCMAKE_TOOLCHAIN_FILE=ios.toolchain.cmake -DPLATFORM=OS64 ../tensorflow/lite`

However I got the following errors:
```
CMake Error at /$(PATH_TO_TENFORFLOW)/tensorflow/ios_cmake_build/flatbuffers/CMakeLists.txt:554 (install):
  install TARGETS given no BUNDLE DESTINATION for MACOSX_BUNDLE executable
  target "flatc".
```
and
```
CMake Error at /$(PATH_TO_TENFORFLOW)/tensorflow/ios_cmake_build/xnnpack/CMakeLists.txt:87 (MESSAGE):
  Unrecognized CMAKE_SYSTEM_NAME = iOS
```

Since you mentioned you've succeeded to build the project on all platforms, have you run into similar errors when building for iOS, and if so, how did you resolve it? Thanks!

**Comment by ebraraktas** [comment]
@zhitaop Sorry for late response. First of all, I don't remember specific options I used at that time when I posted the message. However, I have tried to build with current `master` branch, and managed to build it with the external options below:
```shell
mkdir tf_build && cd tf_build
cmake ../tensorflow_src/tensorflow/lite/ -G Xcode \
  -DCMAKE_TOOLCHAIN_FILE=../ios-cmake/ios.toolchain.cmake \
  -DPLATFORM=OS64 \
  -DFLATBUFFERS_BUILD_FLATC=OFF \
  -DTFLITE_ENABLE_XNNPACK=OFF
cmake --build .
```

Note that disabling XNNPACK may reduce performance, and I don't remember if I could build it with `cmake` for iOS. I added `-DFLATBUFFERS_BUILD_FLATC=OFF` to get rid off `BUNDLE_DESTINATION` error, and could build tflite successfully, but there may exist more proper and elegant solution.
